### PR TITLE
[WIP] LM family support

### PR DESCRIPTION
**Pull request by awygle**
This PR adds support for the LM family of devices. It is a work in progress.

Current status: implemented in icebox.sh and icepack.cc

### CODE REVIEW

File: icebox/icebox.py [hunk 1]
```diff
@@ -1431,6 +1463,16 @@ def parse_db(text, device="1k"):
         (13, 31,  1), #checked
         (19, 31,  2), #checked
     ],
+    "lm4k": [ # TODO(awygle) - learn why the other parts don't match floor planner
```
Comment (daveshah1) [inline icebox/icebox.py]: This probably needs a note somewhere. The last number here is not a cell position, but the global buffer number between 0 and 7.

To work these out, create a global buffer at each location and find out which global number it uses.

File: icebox/icebox.py [hunk 2]
```diff
@@ -1500,6 +1546,12 @@ def parse_db(text, device="1k"):
         "S0":   ( 23, 0, "fabout" ),
         "S1":   ( 24, 0, "fabout" ),
     },
+    "lm4k": {
```
Comment (daveshah1) [inline icebox/icebox.py]: It looks like this is good, but ISTR seeing a Lattice note saying there was no WARMBOOT on the LM series. Maybe that was wrong though

File: icebox/icebox.py [hunk 3]
```diff
@@ -1619,6 +1671,99 @@ def parse_db(text, device="1k"):
...
+        "DIVQ_0":               (11,  0, "PLLCONFIG_3"),
+        "DIVQ_1":               (11,  0, "PLLCONFIG_4"),
+        "DIVQ_2":               (11,  0, "PLLCONFIG_5"),
+        "FILTER_RANGE_0":       (11,  0, "PLLCONFIG_6"),
+        "FILTER_RANGE_1":       (11,  0, "PLLCONFIG_7"),
+        "FILTER_RANGE_2":       (11,  0, "PLLCONFIG_8"),
+        "TEST_MODE":            (12,  0, "PLLCONFIG_8"),
+
+        # PLL Ports
+        # TODO(awygle) confirm these
```
Comment (daveshah1) [inline icebox/icebox.py]: If you haven't already, the way to check these is to create a design using icecube (with pins constrained using a PCF) then run it through icebox_vlog (pass the PCF to this too to get nice net names) and look at where the routing for each signal goes in the output file.

File: icebox/icebox.py [hunk 4]
```diff
@@ -1378,6 +1408,8 @@ def parse_db(text, device="1k"):
         (0, 330, 143): ("padin_glb_netwk", "6"), # (0 0)  (330 143)  (330 143)  routing T_0_0.padin_6 <X> T_0_0.glb_netwk_6
         (0, 331, 143): ("padin_glb_netwk", "7"),
     },
+    "lm4k": {
```
Comment (daveshah1) [inline icebox/icebox.py]: To find these out, create designs using the pad input (SB_GB_IO) for each global input, plus two designs using the internal oscillators if they can drive a global. Then run then through icebox_explain to get the global buffer number, and the ".extra_bit" data is bit position numbers

**Comment by cliffordwolf** [comment]
Very cool!

> Current status: implemented in icebox.sh and icepack.cc

I'd rather not have one ginormous PR for everything LM. So if you feel like the icecube.sh and icepack.cc stuff is finished then I'd say let's make a PR for those and I'd be happy to merge them even when the rest of LM support is not complete yet.

**Comment by awygle** [comment]
Broken up into multiple PRs.

**Pull request closed without merging**